feat: Support OIDC authentication

vishal423 opened the pull request:
Closes #50

vishal423 commented:
@mraible , Would you like to test the Okta integration? I tested once with the developer account and seems good to me.

vishal423 commented:
closes #496

mraible commented:
Yes! I'll test this tomorrow morning.

mraible commented on generators/client/templates/svelte/cypress/integration/account/change-password.spec.js.ejs:
```diff
@@ -1,7 +1,7 @@
 	beforeEach(() => {
 		cy.unregisterServiceWorkers()
-		cy.loginByApi('admin', 'admin')
+		cy.loginByApi(Cypress.env('adminUsername'), Cypress.env('adminPassword'))
```
In the main generator, we allow you to set `CYPRESS_E2E_USERNAME` and `CYPRESS_E2E_PASSWORD` to override the default `admin` credentials. Does this support something similar?

vishal423 replied:
Yes, you can override the credentials with OS environment variables like `CYPRESS_adminUsername`, `CYPRESS_adminPassword` or directly under `cypress.json`.

mraible replied:
I think you should be consistent and use the same variable names that the main generator uses. AFAIK, the main generator never uses the `user` credentials to test anything.

vishal423 replied:
Here, I do have at least [one test](https://github.com/jhipster/generator-jhipster-svelte/runs/2823662541?check_suite_focus=true#step:20:476) to verify that authority as that's the out of box supported user/role. At moment, I believe that test is not included in `OIDC` flow due to a `skip-user-management` condition, however, I plan to fix that.

mraible commented:
I tried using the following JDL:

```
application {
  config {
    baseName blog,
    applicationType monolith
    authenticationType oauth2
    packageName com.jhipster.demo.blog
    prodDatabaseType mysql
    cacheProvider ehcache
    buildTool maven
    clientFramework svelte
    testFrameworks [protractor]
  }
  entities *
}

entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}

entity Post {
  title String required
  content TextBlob required
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{entry}
}

paginate Post, Tag with infinite-scroll
```

I copied this to a `blog-oauth2.jdl` file and ran:

```
jhipster --blueprints svelte jdl blog-oauth2.jdl
```

It fails:

```
Error running generator app: ReferenceError: /Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs:357
    355|   cors:
    356|     # Allow Ionic for JHipster by default (* no longer allowed in Spring Boot 2.4+)
 >> 357|     allowed-origins: "http://localhost:8100,http://localhost:9000<%_ if (!skipClient) { _%>,http://localhost:<%= devServerPort %><%_ } _%>"
    358|     allowed-methods: "*"
    359|     allowed-headers: "*"
    360| <%_ if (authenticationType === 'session') { _%>

devServerPort is not defined
ReferenceError: /Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs:357
    355|   cors:
    356|     # Allow Ionic for JHipster by default (* no longer allowed in Spring Boot 2.4+)
 >> 357|     allowed-origins: "http://localhost:8100,http://localhost:9000<%_ if (!skipClient) { _%>,http://localhost:<%= devServerPort %><%_ } _%>"
    358|     allowed-methods: "*"
    359|     allowed-headers: "*"
    360| <%_ if (authenticationType === 'session') { _%>

devServerPort is not defined
    at module.exports.eval (eval at compile (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:662:12), <anonymous>:374:26)
    at application-dev.yml (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:692:17)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:258:40
    at new Promise (<anonymous>)
    at tryHandleCache (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:256:14)
    at Object.exports.renderFile (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:489:10)
    at Object.renderContent (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/utils.js:286:23)
    at module.exports.template (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base-private.js:758:26)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2222:47
    at Array.map (<anonymous>)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2138:42
    at Array.map (<anonymous>)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2135:29
    at Array.map (<anonymous>)
    at module.exports.writeFilesToDisk (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2134:45)
    at module.exports.writeFiles (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/server/files.js:1730:19) {
  path: '/Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs'
}
ERROR! /Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs:357
    355|   cors:
    356|     # Allow Ionic for JHipster by default (* no longer allowed in Spring Boot 2.4+)
 >> 357|     allowed-origins: "http://localhost:8100,http://localhost:9000<%_ if (!skipClient) { _%>,http://localhost:<%= devServerPort %><%_ } _%>"
    358|     allowed-methods: "*"
    359|     allowed-headers: "*"
    360| <%_ if (authenticationType === 'session') { _%>

devServerPort is not defined
ReferenceError: /Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs:357
    355|   cors:
    356|     # Allow Ionic for JHipster by default (* no longer allowed in Spring Boot 2.4+)
 >> 357|     allowed-origins: "http://localhost:8100,http://localhost:9000<%_ if (!skipClient) { _%>,http://localhost:<%= devServerPort %><%_ } _%>"
    358|     allowed-methods: "*"
    359|     allowed-headers: "*"
    360| <%_ if (authenticationType === 'session') { _%>

devServerPort is not defined
    at module.exports.eval (eval at compile (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:662:12), <anonymous>:374:26)
    at application-dev.yml (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:692:17)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:258:40
    at new Promise (<anonymous>)
    at tryHandleCache (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:256:14)
    at Object.exports.renderFile (/Users/mraible/dev/generator-jhipster-svelte/node_modules/ejs/lib/ejs.js:489:10)
    at Object.renderContent (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/utils.js:286:23)
    at module.exports.template (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base-private.js:758:26)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2222:47
    at Array.map (<anonymous>)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2138:42
    at Array.map (<anonymous>)
    at /Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2135:29
    at Array.map (<anonymous>)
    at module.exports.writeFilesToDisk (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/generator-base.js:2134:45)
    at module.exports.writeFiles (/Users/mraible/dev/generator-jhipster-svelte/node_modules/generator-jhipster/generators/server/files.js:1730:19) {
  path: '/Users/mraible/dev/generator-jhipster/generators/server/templates/src/main/resources/config/application-dev.yml.ejs'
}
Execution time: 12 s.
```

vishal423 commented:
@mraible , Entity generator support is not yet included. For now, you can test with application JDL like used in CI tests - https://github.com/jhipster/generator-jhipster-svelte/blob/9dcff7587081da68fd112eaa85acd234a4258a52/.github/workflows/scripts/monolithic-oidc-maven-postgres-prod.jdl

Also, you can run with `shipster import-jdl <jdl-file>.jdl` to ensure the correct `JHipster` version is included

mraible commented:
I tested with the following JDL and the `shipster` command you recommended:

```
application {
  config {
    baseName SampleBlogApp,
    applicationType monolith,
    authenticationType oauth2,
    packageName tech.jhipster.samples.blog,
    prodDatabaseType postgresql,
    cacheProvider caffeine,
    buildTool maven
    testFrameworks [cypress]
  }
}
```

It generates just fine and I can log in / logout with Keycloak as expected. If I try to run `npm run e2e` with Cypress, it fails.

```
> sample-blog-app@0.1.0 e2e /Users/mraible/svelte-oidc
> cypress run

The plugins file is missing or invalid.

Your `pluginsFile` is set to `/Users/mraible/svelte-oidc/src/test/javascript/cypress/plugins/index.ts`, but either the file is missing, it contains a syntax error, or threw an error when required. The `pluginsFile` must be a `.js`, `.ts`, or `.coffee` file.

Or you might have renamed the extension of your `pluginsFile`. If that's the case, restart the test runner.

Please fix this, or set `pluginsFile` to `false` if a plugins file is not necessary for your project.

 Error: Cannot find module 'puppeteer'
Require stack:
- /Users/mraible/svelte-oidc/src/test/javascript/cypress/plugins/index.ts
- /Users/mraible/Library/Caches/Cypress/7.5.0/Cypress.app/Contents/Resources/app/packages/server/lib/plugins/child/run_plugins.js
```

 I also tried integrating with Okta using `okta apps create jhipster` and sourcing the generated `.okta.env` file. This works too!

I like the support for a dark theme!

vishal423 commented:
I am not sure on how you got `/Users/mraible/svelte-oidc/src/test/javascript/cypress/plugins/index.ts` as we don't generate tests under `src/test/javascript`, also there is no typescript support. Can you try removing `testFrameworks [cypress]` and generate (cypress tests are implicitly included and are not an option)

mraible commented:
I removed `testFrameworks` from the JDL and `npm run e2e` now passes with Keycloak. It is kinda weird to not have Cypress open a browser so you can watch the tests. The main generator opens a browser. 

I tried it with Okta too:

```
$ export CYPRESS_adminUsername=matt.raible
$ export CYPRESS_adminPassword=XXX
$ npm run e2e
```

80% of the tests fail:

```
  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  admin/logger.spec.js                     00:45        6        -        6        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  footer.spec.js                           340ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  home.spec.js                             00:06        3        2        1        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  navbar.spec.js                           00:08        9        5        2        -        2 │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  routes.spec.js                           00:05        3        2        1        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  4 of 5 failed (80%)                      01:06       22       10       10        -        2
```

vishal423 commented:
For now, you can try cypress interactive runner with command `npm run cy:open` and then run one or all specs at once. For failed tests, you can open browser developer tools console window, click on failed assertion/step and see if you can get anything. 

To differentiate between `okta` and `keycloak`, I have included a check on the redirect url to contain `okta`. Do you see that in your okta account?

mraible commented:
> Do you see that in your okta account?

What should I be looking for in my Okta account? The Okta CLI and its `okta apps create jhipster` command will add redirect URIs as follows:

- login: `http://localhost:8080/login/oauth2/code/oidc` and `http://localhost:8761/login/oauth2/code/oidc`
- logout: `http://localhost:8080` and `http://localhost:8761`

vishal423 commented:
I just tried again with a new generated app, changed spring security configurations to point to okta, updated cypress env variables, started app with dev profile, executed cypress tests and everything seems to work on my end.

If you run the cypress tests in interactive runner (with `npm run cy:open`), then, do you see response similar to below for tests under `logger` group? Also, do you have URL similar to highlighted text for your okta account?

![image](https://user-images.githubusercontent.com/15997681/122102653-49b22e00-ce33-11eb-80db-e2421541db90.png)

mraible commented:
If I run `npm run cy:open` and run the loggers test, it doesn't succeed for me.

<img width="1312" alt="Screen Shot 2021-06-15 at 12 36 34 PM" src="https://user-images.githubusercontent.com/17892/122105798-6cfac000-cdd6-11eb-8eef-b003dd36df78.png">

vishal423 commented:
Can you double check your okta credentials that you use in tests?

mraible commented:
I did. They work to log in to the JHipster app in my browser.

<img width="1241" alt="Screen Shot 2021-06-15 at 1 05 17 PM" src="https://user-images.githubusercontent.com/17892/122109442-9f0e2100-cdda-11eb-82d5-65907ab6c738.png">

Just in case I was exporting them wrong, I put them directly into `cypress.json` and I get the same result.

Did you do anything special to set up your Okta app? It usually just requires creating `ROLE_ADMIN` and `ROLE_USER` groups as well as adding a `groups` claim to the ID token.

vishal423 commented:
As you can log in via UI, I don't suspect an issue with your okta setup. 

Previously you mentioned to provide username as `CYPRESS_adminUsername=matt.raible`. Don't you require it to be `matt.raible@okta.com`? If that also doesn't work, then, probably need to check API response to identify root cause.

mraible commented:
Okta doesn't require anything after the `@` for the username. I tried adding `@okta.com` to the admin username in `cypress.json`. The result is the same. I've tried the main generator's Cypress support (with Puppeteer) without `@okta.com` and it works.

vishal423 commented:
Oh. I never tried that way. I just realized that you were able to authenticate successfully via `api/v1/authn`, so, that rules out the issue of wrong credentials. 

Can you check response of the authorize API ? Use `F12` to open browser tools and then click on the authorize url to view request/response details).  Here we programmatically create an authentication session and redirect back to the Spring security authentication callback url. Do you see something similar to below?

![image](https://user-images.githubusercontent.com/15997681/122118748-044b2c00-ce46-11eb-97c7-99e0be300ec8.png)

mraible commented:
The 3rd request fails for me. 

<img width="1783" alt="Screen Shot 2021-06-15 at 3 08 53 PM" src="https://user-images.githubusercontent.com/17892/122124138-1b5d3000-cdec-11eb-81f7-eae152eb60c3.png">

If I look in my system logs on Okta (Reports > System Log), it says there's an "authorization server mismatch". If I drill into the details, I'm not sure if it's trying to hit the default one or `ausqggi8sWC6oHiJb5d6`.

<img width="783" alt="Screen Shot 2021-06-15 at 3 11 57 PM" src="https://user-images.githubusercontent.com/17892/122124481-93c3f100-cdec-11eb-8ae9-23ea7de237ef.png">

It does look like you're using the "org" server for the first request, rather than the default AS. The default AS should have an authorize URL that looks like `https://dev-1309757.okta.com/oauth2/default/v1/authorize`. 

If you grab the issuer from `http://localhost:8080/api/auth-info`, that should be the correct one to use.

vishal423 commented:
Yeah, I was using the org-level association. I have now changed the logic to respect the URL associated with setup (org or specific authorization server) and I believe we should be good now. Let me know if you still see an issue. Thanks

mraible reviewed (approved):
I tested your latest changes. Looks good!

vishal423 commented:
Good to hear. If you have any test credentials (that should pose a low risk even if got accidentally exposed), then, I can set up a CI test to ensure working of OKTA integration.

vishal423 commented:
@pascalgrimaud, Bounty claimed https://opencollective.com/generator-jhipster/expenses/43817

pascalgrimaud commented:
@vishal423 : approved :)))
